core/os/os2/env_linux.odin

```diff
@@ -1,30 +1,202 @@
 //+private
 package os2
 
+import "core:slice"
 import "core:runtime"
+import "core:strings"
+import "core:intrinsics"
+
+// the environment is a 0 delimited list of <key>=<value> strings
+_env_map: map[string]string
```

Comment:
Is there any reason why a `map` is needed in the first place? I know the environment is a delimited `<key>=<value>` array. Does Linux not have something like this already?

Reply:
On Linux, the environment is stored/managed in user space and passed into execve. glibc usually handles all the modification. A map will make for faster lookup, but we could just search through it linearly as well. In my head, I just saw key/value and just immediately thought map. I imagine the only time we would see degradation would be if someone made some kind of `make` clone where environment lookup and modification are more important.

Reply:
I was curious about this point too, so I looked it up. Turns out libc mantains `__environ` variable, which it has gotten from loader where the environment map is stored. When `putenv` or `setenv` is called, it reallocates the buffer for the envs and copies the values. There's also some funny stuff with thread locks on env, because multiple threads may be changing the environment at the same time.

Note that when process is forked, the memory is copied, so the child naturally gets a copy of parent's env. So it doesn't matter where the env is stored as long as both parent and the child access it from the same place.

I personally would love to have our custom env map that is accessible publically, but there's two issues concerning that behaviour:
- Direct insertion into the map is not thread-safe (you might need that if you're planning to spin up children). Honestly not that big of a deal, it just sets up the users for bugs and failures, esp since you don't know when some random-ass library modifies env; but this shouldn't occur because:
- There's no sync with libc's environ. There's no way to specify a custom map type with custom procedure for insertion (at least I don't know of a way to do that, since there's no way to overload operators). The best we can do is to mantain this map with calls to libc's `setenv` and `putenv` when we call `os2` functionality. I don't see performance issues here, since it's a rare operation, but that's the cost of mantaining the map PLUS people would have to make sure not to insert anything into that map directly.

TLDR I didn't read the code so it's just a brain fart for anyone who wants to think about this problem.

Reply:
There is no issue with modifying the map directly because it is private and hidden behind the public API.  It does, however, require a mutex which currently is not in there.  That is "on the list."  I may reverse course and go with a `[]string` that get individually searched.

The idea that we keep track of our own environment is that os2 is supposed to work without `libc`.  If someone is already using `libc.putenv`, then they are probably not then going to use `os2.process_start`, or vice versa, using `os2.put_env` and then calling `libc.system`.  But you raise a good point.  What if a linked library modifies the environment with `libc.putenv` and you want your environments to sync.  I'm not sure how that would work (or if we even care).  If we are going to sync with `libc`, then we should almost just use `libc.putenv/getenv` ourselves.  Maybe we could determine this at compile time with `when NO_CRT` (or whatever it would be called), but that still won't help for third party libraries.